hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/replication/TestGrpcReplicationServiceWithZeroCopy.java

```diff
@@ -0,0 +1,178 @@
+    DatanodeDetails.Port containerPort =
+        DatanodeDetails.newPort(DatanodeDetails.Port.Name.STANDALONE,
+            OzoneConfigKeys.DFS_CONTAINER_IPC_PORT_DEFAULT);
+    DatanodeDetails.Port ratisPort =
+        DatanodeDetails.newPort(DatanodeDetails.Port.Name.RATIS,
+            OzoneConfigKeys.DFS_CONTAINER_RATIS_IPC_PORT_DEFAULT);
+    DatanodeDetails.Port replicationPort =
+        DatanodeDetails.newPort(DatanodeDetails.Port.Name.REPLICATION,
+            replicationConfig.getPort());
+    DatanodeDetails.Port streamPort =
+        DatanodeDetails.newPort(DatanodeDetails.Port.Name.RATIS_DATASTREAM,
+            OzoneConfigKeys.DFS_CONTAINER_RATIS_DATASTREAM_PORT_DEFAULT);
+    dn.addPort(containerPort);
+    dn.addPort(ratisPort);
+    dn.addPort(replicationPort);
+    dn.addPort(streamPort);
+
+    datanode = dn.build();
+
+    final String testDir =
+        Files.createDirectory(tempDir.resolve("VolumeDir")).toString();
+
+    MutableVolumeSet volumeSet = mock(MutableVolumeSet.class);
+    when(volumeSet.getVolumesList()).thenReturn(
+        Collections.singletonList(new HddsVolume.Builder(testDir).conf(conf).build()));
+
+    Handler containerHandler = mock(Handler.class);
+    ContainerController containerController =
+        new ContainerController(containerSet,
+            Collections.singletonMap(ContainerProtos.ContainerType.KeyValueContainer,
+                containerHandler));
+
+    ContainerImporter importer =
+        new ContainerImporter(conf, containerSet, containerController,
+            volumeSet);
+
+    replicationServer =
+        new ReplicationServer(containerController, replicationConfig, secConf,
+            null,
+            importer, datanode.threadNamePrefix());
+    replicationServer.start();
+  }
+
+  @AfterEach
+  public void cleanup() {
+    replicationServer.stop();
+  }
+
+  @Test
+  public void testZeroCopyDownload() throws IOException {
+    SimpleContainerDownloader downloader =
+        new SimpleContainerDownloader(conf, null);
+    Path downloadDir = Files.createDirectory(tempDir.resolve("DownloadDir"));
+    Path result = downloader.getContainerDataFromReplicas(containerId,
+        Collections.singletonList(datanode), downloadDir,
+        CopyContainerCompression.NO_COMPRESSION);
+
+    assertTrue(result.toString().startsWith(downloadDir.toString()));
+
+    File[] files = downloadDir.toFile().listFiles();
+
+    assertNotNull(files);
+    assertEquals(files.length, 1);
+
+    assertTrue(files[0].getName().startsWith("container-" + containerId + "-"));
+
+    downloader.close();
+  }
+
+  @Test
+  public void testZeroCopyUpload() throws IOException {
+    GrpcContainerUploader uploader =
+        new GrpcContainerUploader(conf, null);
+
+    CompletableFuture<Void> callback = new CompletableFuture<>();
+    OutputStream out = uploader.startUpload(containerId,
+        datanode, callback,
+        CopyContainerCompression.NO_COMPRESSION);
+    out.close();
```

Comment:
Shouldn't we check the upload destination to verify the file was transferred correctly?